File: configure.rb
#!/usr/bin/env ruby
s = <<-SH
CC=g++ -std=c++11 -Wall -Llib -Iinclude -lstdc++
EXFLAGS=-Wl,-rpath=./lib
SOFLAGS=-shared -fPIC

_all: all
bin:
	mkdir bin
lib:
	mkdir lib

SH

srcfiles = Dir['*.cpp'].map {|filename| filename[0...-4]}

binsrc = srcfiles.select {|filename| filename.match(/[0-9]{3}/)}
libsrc = srcfiles.reject {|filename| filename.match(/[0-9]{3}/)}

binfiles = binsrc.map {|filename| 'bin/' + filename.match(/0*(.*)/)[1]}
libfiles = libsrc.map {|filename| 'lib/lib' + filename.match(/([^0-9]*)/)[1] + '.so'}

s << "all: " + binfiles.join(' ') + "\n\n"
s << "clean:\n\trm -r bin lib\n\n"

binfiles.each_with_index do |filename, i|
  deps = []
  File.open(binsrc[i] + ".cpp", "r").each_line do |line|
    if line.start_with?('#include "')
      deps << line.match(/"(.*)\.h"/)[1]
    end
  end
  s << filename + ": " + binsrc[i] + ".cpp "
  s << deps.map {|dep| "lib/lib" + dep + ".so"} .join(' ')
  s << "| bin\n\t"
  s << '$(CC) $(EXFLAGS) ' + binsrc[i] + '.cpp -o ' + filename + ' '
  s << deps.map {|dep| '-l' + dep} .join(' ')
  s << "\n\n"
end

libfiles.each_with_index do |filename, i|
  deps = []
  File.open(libsrc[i] + ".cpp", "r").each_line do |line|
    if line.start_with?('#include "')
      dep = line.match(/"(.*)\.h"/)[1]
      deps << dep unless dep == libsrc[i]
    end
  end
  s << filename + ": " + libsrc[i] + ".cpp "
  s << deps.map {|dep| "lib/lib" + dep + ".so"} .join(' ')
  s << "| lib\n\t"
  s << '$(CC) $(SOFLAGS) ' + libsrc[i] + '.cpp -o ' + filename + ' '
  s << deps.map {|dep| '-l' + dep} .join(' ')
  s << "\n\n"
end

File.open('makefile', 'w').write(s)

File: configure.py
#!/usr/bin/env python
import os, re
EXCLUDE_FILE_EXTENSIONS = (".swp",)

s = """\
CC=gcc
CXX=g++ -std=c++17 -lstdc++
CFLAGS+= -g -Wall -Werror -Llib -Iinclude
EXFLAGS=-Wl,-rpath=./lib
SOFLAGS=-shared -fPIC
.PRECIOUS: build/%.o

# Patterns
all: _all
clean:
	rm -rf ./bin ./lib ./build

test:
	python -c 'import unittest; from test import *; unittest.main()'

bin/ lib/ build/:
	mkdir -p $@

build/%.o: src/%.cpp | build/
	$(CXX) $(CFLAGS) $(EXFLAGS) $< -c -o $@

bin/%: src/%.cpp | bin/
	$(CXX) $(CFLAGS) $(EXFLAGS) $(if $(findstring lib/lib,$^),\\
		$(filter-out $<,$(patsubst lib/lib%.so,-l%,$^)),\\
		$<) -o $@

lib/lib%.so: src/%.cpp | lib/
	$(CXX) $(CFLAGS) $(SOFLAGS) $(patsubst lib/lib%.so,-l%,$^) -o $@

# Dependencies
"""

srcfiles = sorted(os.listdir('src/'))
binsrc, binfiles, libfiles = [], [], {}
for f in srcfiles:
    if re.search(".*(\..*)", f)[1] in EXCLUDE_FILE_EXTENSIONS:
        continue
    if re.match('[0-9]{3}', f):
        binsrc.append('src/' + f)
        binfiles.append('bin/' + re.sub('\..+', '', f))
    else:
        k = re.sub('\..*', '', f)
        libfiles[k] = ('src/' + f, 'lib/lib' + k + '.so')

s += "_all: " + ' '.join(binfiles) + "\n\n"

for srcname, binname in zip(binsrc, binfiles):
    deps = []
    with open(srcname) as fo:
        for line in fo:
            match = re.search('#include "(.*)\.h"', line)
            if match:
                deps.append(match[1])
    if not deps:
        continue
    srcname = re.sub('src/(.*)\.cpp', 'build/\\1.o', srcname)
    s += binname + ": " + srcname + ' '
    s += ' '.join(libfiles[dep][1] for dep in deps) + '\n'
s += '\n'

for k in libfiles:
    srcname, libname = libfiles[k]
    deps = []
    with open(srcname) as fo:
        for line in fo:
            match = re.search('#include "(.*)\.h"', line)
            if match and match[1] != k:
                deps.append(match[1])
    if not deps:
        continue
    s += libname + ": " + srcname + ' '
    s += ' '.join(libfiles[dep][1] for dep in deps) + '\n'

open('makefile', 'w').write(s)
